ocaml/xapi/xapi_event.ml

```diff
@@ -541,159 +594,118 @@ let from_inner __context session subs from from_t timer batching =
+          Throttle.Batching.with_recursive_loop batching @@ fun self since ->
+          let result =
+            Db_lock.with_lock (fun () -> grab_range ~since (Db_backend.make ()))
           in
+          let msg_gen, messages, _tables, events = result in
+          let {creates; mods; deletes; last} = events in
           if
             creates = []
             && mods = []
             && deletes = []
             && messages = []
             && not (Clock.Timer.has_expired timer)
           then (
-            last_generation := last ;
-            (* Cur_id was bumped, but nothing relevent fell out of the db. Therefore the *)
+            (* cur_id was bumped, but nothing relevent fell out of the database.
+               Therefore the last ID the client got is equivalent to the current one. *)
             sub.cur_id <- last ;
-            (* last id the client got is equivalent to the current one *)
             last_msg_gen := msg_gen ;
             wait2 sub last timer ;
-            (self [@tailcall]) arg
+            (* The next iteration will fold over events starting after
+               the last database event that matched a subscription. *)
+            let next = last in
+            (self [@tailcall]) next
           ) else
             result
         in
-        grab_nonempty_range ()
+        grab_nonempty_range from
     )
   in
-  last_generation := last ;
-  let event_of op ?snapshot (table, objref, time) =
+  let {creates; mods; deletes; last} = events in
+  let event_of op ?snapshot {table; obj; time} =
     {
       id= Int64.to_string time
     ; ts= "0.0"
     ; ty= String.lowercase_ascii table
     ; op
-    ; reference= objref
+    ; reference= obj
     ; snapshot
     }
   in
-  let events =
-    List.fold_left
-      (fun acc x ->
-        let ev = event_of `del x in
-        if Subscription.event_matches subs ev then ev :: acc else acc
-      )
-      [] deletes
-  in
-  let events =
-    List.fold_left
-      (fun acc (table, objref, mtime) ->
-        let serialiser = Eventgen.find_get_record table in
-        try
-          let xml = serialiser ~__context ~self:objref () in
-          let ev = event_of `_mod ?snapshot:xml (table, objref, mtime) in
-          if Subscription.event_matches subs ev then ev :: acc else acc
-        with _ -> acc
-      )
-      events mods
+  let events_of ~kind ?(with_snapshot = true) entries acc =
+    let rec go events ({table; obj; time= _} as entry) =
+      try
+        let snapshot =
+          let serialiser = Eventgen.find_get_record table in
+          if with_snapshot then
+            serialiser ~__context ~self:obj ()
+          else
+            None
+        in
+        let event = event_of kind ?snapshot entry in
+        if Subscription.event_matches subs event then
+          event :: events
+        else
+          events
+      with _ ->
+        (* CA-91931: An exception may be raised here if an object's
```

Comment:
Would it be safe to match specific exceptions to avoid swallowing unexpected exceptions?